core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/async/SyncResponseHandlerAdapter.java

```diff
@@ -76,7 +76,7 @@ public void onError(Throwable err) {
     @Override
     public CompletableFuture<T> prepare() {
         streamFuture = new CompletableFuture<>();
-        return streamFuture.thenComposeAsync(baos -> {
+        return streamFuture.thenCompose(baos -> {
```

Comment:
Hmm the stream future is completed by thread calling `onComplete` on the subscriber which will be owned by the async HTTP client. If we complete on the same thread we would block the HTTP client's thread while `responseHandler` runs. Should we use the future completion executor?

Reply:
The `responseHandler` should run fairly fast. I wonder if it's worth delegating to the future completion executor.

Reply:
That would depend on the size and complexity of the response I suppose. A DDB `BatchGetItem` could return up to 16MB and could take a non-negligible time to un marshall for example (though I've never measured it).

Reply:
Yeah, ec2 response unmarshalling could also take non-negligible time. I'd be interested to see which is more costly, unmarshalling large payload or transitioning task from one thread to another. Need to run some benchmarks.

Looking at page 63 of the CompletableFuture perf presentation slides https://qconsf.com/sf2017/system/files/presentation-slides/cf.pdf, seems like offloading executor would add large overhead for small quick task and small overhead for large tasks that take 100ms.

Reply:
Cool. I don't think this needs to be a blocker. We can go with a synchronous compose for now and offload if we need to (or provide the option); I think large responses are the exception not the norm.

Reply:
Something CPU-bound _should_ be fine to happen on the event loop threads. It's a decent bit of overhead to pass to another thread.

Reply:
@millems not sure. Ideally the event loop threads should only be handling I/O IMO. For example, if there was only 1 event loop thread, and multiple concurrent requests it shouldn't be blocked too long doing non I/O things for any one request, especially if data is waiting to be sent/received.

Reply:
Yeah, it's possible. Benchmarking will tell us the true story.